### DISCUSSION TITLE
How to include constraints in custom differential action models?

### DISCUSSION THREAD

**Discussion by lwygzh** [loco-3d/crocoddyl#1205] [Q&A]
Hi,

I wanted to create a custom differential action model that has constraints. I was wondering if there is any sample code showing how to do that, especially when taking in a constraint manager instance instead of manually specifying the partial derivatives and passing it to `data`.

Thanks!

**Comment by cmastalli** [comment]
We're refactoring deeply Crocoddyl. For instance, we'll deprecate differentiable action models in future releases. This constraint us to develop the requested example.

However, I suggest you have a look at our inverse-dynamics differential action models. In a nutshell, the constraint manager works similarly to our cost sum, i.e., you need to share the memory: https://github.com/loco-3d/crocoddyl/blob/devel/include/crocoddyl/multibody/actions/free-invdyn.hpp#L401

**Reply by lwygzh** [reply]
@cmastalli  Thanks for the response! I don't think I have access to the file you shared.

**Reply by cmastalli** [reply]
I updated the link.
